### DISCUSSION TITLE
Passing models as parameters to a function with structural typing

### DISCUSSION THREAD

**Discussion by rhz** [prisma/prisma#25130] [Q&A]
### Question

The question is how can I pass a model name or delegate to a function and make use of its method in a generic way. I believe this question has been raised in this forum before in different forms and with different goals in mind. Here is, I think, a perhaps a slightly different way to think about it.

TypeScript has structural typing, so in the type signature of a function we say something like the function can be passed any object that has a propery named `length` and is of type `number`. For example,
```ts
const f = (x: { length: number }) => x.length+2
```
Then the type checker verifies that we in fact only access that propery on the object and we use it according to its type. *Can we do the same in Prisma? Can we write a function that receives any model that has a field named `createdAt` of type `DateTime` and then have the type checker verify we only access that field and we use its value in according to its type?*

The motivation comes from the following code I have in a TypeScript + Express + Prisma web service I'm writting.
```ts
app.get('/status', async (req: Request, res: Response) => {
  res.status(200).send({
    lastReceived: {
      'user.signup': (await prisma.userSignup.findFirst({
        orderBy: { createdAt: 'desc' },
        select: { createdAt: true }
      }))?.createdAt.toString(),
      'enrollment.created': (await prisma.enrollment.findFirst({
        orderBy: { createdAt: 'desc' },
        select: { createdAt: true }
      }))?.createdAt.toString(),
      'enrollment.progress': (await prisma.enrollmentProgress.findFirst({
        orderBy: { updatedAt: 'desc' },
        select: { updatedAt: true }
      }))?.updatedAt.toString(),
      'order.created': (await prisma.order.findFirst({
        orderBy: { createdAt: 'desc' },
        select: { createdAt: true }
      }))?.createdAt.toString(),
    }
  })
})
```
That's a lot of repetition and there might be more models to come. I tried abstracting these queries into a function, but it turns out it's way more difficult than I expected. The first attempt at this was inspired by a [discussion on this repo](
https://github.com/prisma/prisma/discussions/10580#discussioncomment-8180884).
```typescript
const getLast = async (modelName: Prisma.ModelName, fieldName: string) =>
  (await prisma[modelName].findFirst({
    orderBy: { [fieldName]: 'desc' },
    select: { [fieldName]: true }
  }))?.[fieldName].toString()
```
but I get the error
```
src/index.ts:320:19 - error TS2551: Property 'UserSignup' does not exist on type 'PrismaClient<PrismaClientOptions, never, DefaultArgs>'. Did you mean 'userSignup'?
```
That's because `Prisma.ModelName` has model names in PascalCase and `prisma[modelName]` expects model names in camelCase. Then I tried
```typescript
const getLast = async <M extends Prisma.ModelName>(
  model: PrismaClient[Uncapitalize<M>],
  fieldName: string)
  : Promise<string> =>
  (await model.findFirst({
    orderBy: { [fieldName]: 'desc' },
    select: { [fieldName]: true }
  }))?.[fieldName].toString()
```
and I got the error
```
src/index.ts:328:18 - error TS2349: This expression is not callable.
  Each member of the union type '(<T extends UserSignupFindFirstArgs>(args?: SelectSubset<T, UserSignupFindFirstArgs<DefaultArgs>> | undefined) => Prisma__UserSignupClient<GetFindResult<...> | null, null, DefaultArgs>) | ... 6 more ... | (<T extends ItemFindFirstArgs>(args?: SelectSubset<...> | undefined) => Prisma__ItemClient<...>)' has signatures, but none of those signatures are compatible with each other.

328     (await model.findFirst({
                     ~~~~~~~~~
```
That felt like I was very close. If only I could convince the compiler that indeed the query was generic enough that it'd work on the `model` and `fieldName` that I'm passing to `getLast`.

After that I decided to work on a smaller problem. Let's fix the model, because that part seems really difficult, and just pass the `fieldName` to the function. So I tried the following and it works!
```typescript
 const getLast = (fieldName: Prisma.UserSignupScalarFieldEnum) =>
    prisma.userSignup.findFirst({
      orderBy: { [fieldName]: 'desc' },
      select: { [fieldName]: true }
    }).then((result) => result?.[fieldName])
```
After digging some more into the auto-generated Prisma client code (in `node_modules/.prisma/client/index.d.ts`) I found out that `keyof Prisma.UserSignupFieldRefs` can be used in place of `Prisma.UserSignupScalarFieldEnum` for the type of `fieldName`. That's fortunate, because in `Prisma.TypeMap` there's a map
from the model name to the `...FieldRefs`. With this knowledge I can rewrite the function above into
```typescript
const f = (
  modelName: Uncapitalize<Prisma.ModelName>,
  fieldName: keyof (Prisma.TypeMap["model"][
    Capitalize<typeof modelName>]["fields"])) =>
  prisma[modelName].findFirst({
    orderBy: { [fieldName]: 'desc' },
    select: { [fieldName]: true }
  }).then((result) => result?.[fieldName])
```
which still gives me the same error. That feels like it's got really really close. Given that `fieldName` is guaranteed to exist in the model, what information is the type checker missing to accept this function?

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: EndeavourOS
- Database: SQLite
- Node.js version: v21.6.2
- Prisma version: 5.18.0

**Comment by ludralph** [comment]
Hi @rhz 👋 

Can you please try out this code?
```js
import { PrismaClient, Prisma } from '@prisma/client';
import express, { Request, Response } from 'express';

const prisma = new PrismaClient();
const app = express();

// Utility function to convert PascalCase to camelCase
const toCamelCase = (str: string) => str.charAt(0).toLowerCase() + str.slice(1);

// Generic function to get the last record based on a field
const getLast = async <M extends Prisma.ModelName>(
  modelName: M,
  fieldName: string
): Promise<string | undefined> => {
  const model = prisma[toCamelCase(modelName) as keyof PrismaClient];
  const result = await (model as any).findFirst({
    orderBy: { [fieldName]: 'desc' },
    select: { [fieldName]: true }
  });
  return result?.[fieldName]?.toString();
};

app.get('/status', async (req: Request, res: Response) => {
  res.status(200).send({
    lastReceived: {
      'user.signup': await getLast('UserSignup', 'createdAt'),
      'enrollment.created': await getLast('Enrollment', 'createdAt'),
      'enrollment.progress': await getLast('EnrollmentProgress', 'updatedAt'),
      'order.created': await getLast('Order', 'createdAt'),
    }
  });
});
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
